### PR TITLE
Make estimation error visible

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- The estimation error was hidden. Now it's visible.
- Issue originally reported on Slack ([link](https://spaceshard.slack.com/archives/C03L03VF7PB/p1699797003112339))

## Development related changes

- Key changes done in `crates/starknet/src/starknet/estimations.rs`:
  - We were passing an empty error reason vector
  - But even if the vector was populated, the reason was concealed by the blockifier error wrapping it (`blockifier::transaction::errors::TransactionExecutionError::ExecutionError`)
- Introduces a new Cairo contract for testing
  - I needed a contract that can fail (panic), and even do so with a custom reason
  - Source included for easier future modifications
    - Related:
      - #124
      - #37

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Checked the TODO section in README.md if this PR resolves it
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-server/tests/test_estimate_fee.rs [hunk 1]
```diff
@@ -354,6 +356,70 @@ mod estimate_fee_tests {
...
+        let account = Arc::new(SingleOwnerAccount::new(
+            devnet.clone_provider(),
+            signer,
+            account_address,
+            CHAIN_ID,
+            ExecutionEncoding::Legacy,
+        ));
+
+        // get class
+        let contract_artifact: SierraClass = load_json(CAIRO_1_BALANCE_CONTRACT_SIERRA_PATH);
```
Comment (mikiw) [inline crates/starknet-server/tests/test_estimate_fee.rs:374]: so there is a `get_events_contract_in_sierra_and_compiled_class_hash()` function and I refactor it to `get_contract_in_sierra_and_compiled_class_hash()` https://github.com/0xSpaceShard/starknet-devnet-rs/commit/1f5e010254659c50688baa27fb709369846a399b maybe this code can be replaced with that

File: crates/starknet-server/test_data/cairo1/balance_contract/balance_contract.cairo [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#[starknet::interface]
+trait IBalanceContract<TContractState> {
+   fn increase_balance(ref self: TContractState, amount1: u128, amount2: u128);
```
Comment (mikiw) [inline crates/starknet-server/test_data/cairo1/balance_contract/balance_contract.cairo:3]: Do we need increase_balance and get_balance in this contract I have seen that only create_panic is called?

File: crates/starknet-server/tests/test_estimate_fee.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet-server/tests/test_estimate_fee.rs]: I changed the name, but the implementation is (almost) the same: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/247/commits/5d177fb6e861be028865fa4373184f7676c5760e
I applied it in other places in my code too

File: crates/starknet-server/test_data/cairo1/balance_contract/balance_contract.cairo (hunk 2 above)
Comment (FabijanC) [reply crates/starknet-server/test_data/cairo1/balance_contract/balance_contract.cairo]: I minimized the contract used and renamed it accordingly: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/247/commits/f4d199d4b6e74a6c29ffce1ece470a89e47ee26b

**Review by mikiw — APPROVED** [review]

File: crates/starknet-server/tests/common/utils.rs [hunk 3]
```diff
@@ -44,6 +44,20 @@ pub fn load_json<T: serde::de::DeserializeOwned>(path: &str) -> T {
     loaded
 }
 
+pub fn get_flattened_sierra_contract_and_casm_hash(
```
Comment (marioiordanov) [inline crates/starknet-server/tests/common/utils.rs:47]: function `get_flattened_sierra_contract_and_casm_hash` is identical to `get_events_contract_in_sierra_and_compiled_class_hash` why not leaving only one of them, preferably `get_flattened_sierra_contract_and_casm_hash`

Comment (FabijanC) [reply crates/starknet-server/tests/common/utils.rs:47]: Reused the function in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/247/commits/cd8269cce80110f608f0bc5ad7973ca562d7bcd5

**Pull request merged**